augur/filter/subsample.py

```diff
@@ -72,7 +72,7 @@ def get_groups_for_subsampling(strains, metadata, group_by=None):
     metadata = metadata.loc[list(strains)]
     group_by_strain = {}
 
-    if metadata.empty:
+    if len(metadata) == 0:
```

Comment:
Surprised this bug fix isn't reflected in tests that break!

Reply:
Huh, but I guess it's hard to get a single column metadata file into `augur filter`?

```
$ augur filter --metadata <(printf '%s\n' strain a b c) --output-metadata /dev/stdout
ERROR: Could not determine the delimiter of '/dev/fd/63'. Valid delimiters are: (',', '\t'). This can be changed with --metadata-delimiters.
```

Based on the error message, I thought I could force it by adding `--metadata-delimiters ','`, but that doesn't work because it just reduces the set of inferred delimiters; it doesn't set the delimiter.

Reply:
I don't remember how I discovered this... metadata shouldn't often be used with no columns (especially for subsampling). I'll add a doctest anyways.

Reply:
I remember now. Prior to merging 74e7fe91949c546e0878b26013764c38c50f6a72, I had tried something simple like this Cram test:

```
  $ cat >metadata.tsv <<~~
  > strain
  > SEQ_1
  > SEQ_2
  > SEQ_3
  > SEQ_4
  > SEQ_5
  > ~~

  $ ${AUGUR} filter \
  >  --metadata metadata.tsv \
  >  --subsample-max-sequences 2 \
  >  --subsample-seed 0 \
  >  --output-strains filtered_strains.txt
```

Without the change in this PR (b89694cb418393333f3163a0b59302686fe82d79), `filtered_strains.txt` was an empty file.

With the change, it did what I expected:

```
  $ cat filtered_strains.txt
  SEQ_1
  SEQ_5
```

However, 74e7fe91949c546e0878b26013764c38c50f6a72 prevents using any metadata without columns since no delimiter is detected on a column-less file.

Since it has no practical use anymore, I'm going to drop b89694cb418393333f3163a0b59302686fe82d79 from this PR.

Reply:
Oops, didn't see your 2nd comment until just now when I refreshed the page. Looks like we reached the same conclusion.

Reply:
Ah, I would have kept it. :-) I still think it's a valid correction, even if it's not a bug that can be run into any more given other restrictions. Those restrictions may change again in the future, revealing the latent bug.

> However, [74e7fe9](https://github.com/nextstrain/augur/commit/74e7fe91949c546e0878b26013764c38c50f6a72) prevents using any metadata without columns since no delimiter is detected on a column-less file.

Hmm. I'd say our example metadata files with a `strain` column are single-column metadata, not column-less...

Reply:
> Those restrictions may change again in the future, revealing the latent bug.

Good point. I'll add it back with a doctest in another PR.

> Hmm. I'd say our example metadata files with a `strain` column are single-column metadata, not column-less...

Right. I was thinking in terms of the DataFrame representation which removes the `strain` column and uses it as the DataFrame index, thus making it appear "column-less". This isn't ideal and subcommands such as `export v2` add the column back [explicitly](https://github.com/nextstrain/augur/blob/4bf72f8a257e60a8f75ec327dc2850950af2968e/augur/export_v2.py#L1120-L1121). It's not trivial to retain the column in `read_metadata()` which (potentially) returns a chunk iterator, so I think we'll have to live with it for usages of `read_metadata()`.

Reply:
Adding back in #1408